sympy/core/expr.py

```diff
@@ -684,7 +684,9 @@ def is_constant(self, *wrt, **flags):
                     if b is S.NaN:
                         # evaluation may succeed when substitution fails
                         b = expr._random(None, 1, 0, 1, 0)
-                except ZeroDivisionError:
+                except (ZeroDivisionError, ValueError):
+                    # In case of (at least) a Relational, a division by zero
+                    # will show up here as a ValueError
```

Comment:
I think it's problematic that Relationals are treated as expressions in so many places. What does `is_constant` mean for a Relational? Tracing that back to where it gets called does it actually make sense to test if the Relational is constant?

Reply:
No idea. But if you look at the `expr` test it also triggers the bug, so it is not really Relationals that are at fault here, but `equals`. (Or rather that `zoo` is returned for division by zero and that `zoo` is not comparable, so it cannot be an argument to `Max` or `Min`.)

Reply:
So the comment is a bit misleading once I understood a bit more...

Reply:
So is this not from calling Relational.is_constant?

Reply:
Well, when calling `simplify` it calls `equals`, which calls `is_constant`. The problem is that it tests replacing the variables with 1 which is not very good for `Max(y/(1-x), ...)` as `zoo` is returned which cannot be an argument to `Max` (as opposed to earlier, when it was `oo`). Hence, the comment is actually not incorrect ("at least"), but maybe not spot on either.

Reply:
I'm wondering if one actually should remove the ZeroDivisionError exception? These do not seem to happen anymore...

Reply:
There are still ways to get ZeroDivisionError:
```julia
In [3]: (x % y).subs(y, 0)                                                                                                        
---------------------------------------------------------------------------
ZeroDivisionError
```
I think `(x % (y-1)).is_constant()` would reach this line.

I do think exceptions need cleaning up though. `x % 0` could just as easily be nan.

Reply:
It also happens when dividing with an empty polynomial.

I introduced an `InvalidComparison` exception which I think would be a good way since one can tell that it is a comparison that went wrong (also Min/Max raised ValueError, not TypeError, so catching two such big classes is probably not a good idea...)

Reply:
I just tried unsuccessfully. How do I get poly to raise ZeroDivisionError?
```julia
In [11]: Poly([], x) / Poly([], x)                                              
Out[11]: nan

In [12]: Poly(x, x) / Poly([], x)                                               
Out[12]: zoo⋅x

In [13]: Poly(x**2-1, x) / Poly([], x)                                          
Out[13]: 
    ⎛ 2    ⎞
zoo⋅⎝x  - 1⎠

In [14]: Poly(x**2-1, x) / Poly(0, x)                                           
Out[14]: 
    ⎛ 2    ⎞
zoo⋅⎝x  - 1⎠
```
It happens when calling the div method:
```julia
In [16]: Poly(x, x).div(Poly(0, x))                                             
---------------------------------------------------------------------------
ZeroDivisionError 
```
But I don't see how to get that from e.g. subs.

Reply:
Looks like `Poly.__div__` converts to `Expr` before dividing...

Reply:
I just searched for instances of `ZeroDivisionError`... So not sure how to get that in practice.

While `InvalidComparison` works nicely for comparisons and Min/Max, there are other methods that can only accept a subset of the numerical range as input. There will be errors for these for certain `is_constant` expressions. Not sure if all raises `ValueError`, but it seems to make sense to catch that as well.

Reply:
More broadly we should try and come up with better exceptions to raise in different situations. Far too much library code catches very generic exceptions.

I removed most examples catching AttributeError in #16164. I just checked now and 6 new ones appeared in the time since. There are over 100 lines catching TypeError and same for ValueError which is far too much scope for masking bugs. Part of the reason this is bad practice though is that it is difficult to understand what the code does and why - in itself that makes it hard to remove each example.